### PR TITLE
parallelize invalidateRepositories()

### LINKED ISSUE #1353: HistoryGuru.invalidateRepositories() is too slow
When the webapp is being started, it executes all kinds of SCM commands, to get tags etc. This takes significant time with tens of repositories. The stack trace looks like this:

```
 ffff80ffaa7580ac * org/opensolaris/opengrok/history/MercurialRepository.isWorking()Z+28 (line 1118)
 ffff80ffaac93c0c * org/opensolaris/opengrok/history/HistoryGuru.invalidateRepositories(Ljava/util/Collection;)V+103 (line 1540)
 ffff80ffaa2082bd * org/opensolaris/opengrok/configuration/RuntimeEnvironment.setConfiguration(Lorg/opensolaris/opengrok/configuration/Configuration;)V+49 (line 2458)
 ffff80ffaa2082bd * org/opensolaris/opengrok/configuration/RuntimeEnvironment.readConfiguration(Ljava/io/File;)V+5 (line 2254)
 ffff80ffaa2082bd * org/opensolaris/opengrok/web/WebappListener.contextInitialized(Ljavax/servlet/ServletContextEvent;)V+45 (line 140)
 ffff80ffaa208302 * org/apache/catalina/core/StandardContext.listenerStart()Z+628 (line 9476)
 ffff80ffaa207a90 * org/apache/catalina/core/StandardContext.startInternal()V+1132 (line 10189)
 ffff80ffaa2082bd * org/apache/catalina/util/LifecycleBase.start()V+210 (line 270)
 ffff80ffaa208302 * org/apache/catalina/core/ContainerBase.addChildInternal(Lorg/apache/catalina/Container;)V+170 (line 1422)
 ffff80ffaa2082bd * org/apache/catalina/core/ContainerBase.addChild(Lorg/apache/catalina/Container;)V+26 (line 1395)
 ffff80ffaa2082bd * org/apache/catalina/core/StandardHost.addChild(Lorg/apache/catalina/Container;)V+40 (line 1463)
 ffff80ffaa208302 * org/apache/catalina/startup/HostConfig.deployWAR(Lorg/apache/catalina/util/ContextName;Ljava/io/File;)V+1892 (line 1968)
 ffff80ffaa2082bd * org/apache/catalina/startup/HostConfig$DeployWar.run()V+12 (line 3646)
 ffff80ffaa208302 * java/util/concurrent/Executors$RunnableAdapter.call()Ljava/lang/Object;+4 (line 1022)
 ffff80ffaa208114 * java/util/concurrent/FutureTask.run()V+42 (line 520)
 ffff80ffaa208302 * java/util/concurrent/ThreadPoolExecutor.runWorker(Ljava/util/concurrent/ThreadPoolExecutor$Worker;)V+95 (line 2263)
 ffff80ffaa2082bd * java/util/concurrent/ThreadPoolExecutor$Worker.run()V+5 (line 1234)
 ffff80ffaa208302 * java/lang/Thread.run()V+11 (line 1489)
 ffff80ffaa2007a7 * java/lang/Thread.run()V+-10632 (line 1491)
 ffff80ffba3e0c1f __1cJJavaCallsLcall_helper6FpnJJavaValue_pnMmethodHandle_pnRJavaCallArguments_pnGThread__v_ () + 8d7
 ffff80ffba3dc85c __1cJJavaCallsMcall_virtual6FpnJJavaValue_nLKlassHandle_pnGSymbol_5pnRJavaCallArguments_pnGThread__v_ () + 424
 ffff80ffba3dcc44 __1cJJavaCallsMcall_virtual6FpnJJavaValue_nGHandle_nLKlassHandle_pnGSymbol_6pnGThread__v_ () + 60
 ffff80ffba55fe2c __1cMthread_entry6FpnKJavaThread_pnGThread__v_ () + b8
 ffff80ffbaaf94dd __1cKJavaThreadDrun6M_v_ () + 5e1
 ffff80ffba8dd7a5 java_start () + 175
 ffff80ffbf5f94a1 _thrp_setup () + a5
 ffff80ffbf5f9780 _lwp_start ()
```

Possibly the sequential nature of `invalidateRepositories()` could be converted to parallel.

### PR DESCRIPTION
**Pull request by vladak**
fixes #1353

### CODE REVIEW

**Comment by vladak** [comment]
On Solaris kernel zone with local ZFS backed storage indexing 700+ git repositories the invalidation step was reduced from 17 to 4 seconds. Looks like for each repo `git --help` and `git log -1 ...` commands were executed.

File: src/org/opensolaris/opengrok/history/HistoryGuru.java [hunk 1]
```diff
@@ -36,6 +36,8 @@
 import java.util.Set;
 import java.util.concurrent.CountDownLatch;
 import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
```
Comment (tulinkry) [inline src/org/opensolaris/opengrok/history/HistoryGuru.java]: copyright

File: src/org/opensolaris/opengrok/history/RepositoryFactory.java [hunk 2]
```diff
@@ -82,9 +82,9 @@ private RepositoryFactory() {
      * @param file File that might contain a repository
      * @return Correct repository for the given file
      * @throws java.lang.InstantiationException in case we cannot create the
-     * repo object
```
Comment (tulinkry) [inline src/org/opensolaris/opengrok/history/RepositoryFactory.java]: copyright

**Comment by tulinkry** [comment]
This history invalidation runs before the index so the threads won't fight with the indexer? Just to be sure.

File: src/org/opensolaris/opengrok/history/HistoryGuru.java (hunk 1 above)
Comment (vladak) [inline src/org/opensolaris/opengrok/history/HistoryGuru.java]: fixed

File: src/org/opensolaris/opengrok/history/RepositoryFactory.java (hunk 2 above)
Comment (vladak) [inline src/org/opensolaris/opengrok/history/RepositoryFactory.java]: fixed

**Comment by vladak** [comment]
invalidateRepositories is called in:
  - setConfiguration() which is called from Indexer's main() before the call to prepareIndexer()
  - addRepositories() called from prepareIndexer() before it starts the real work (when scanning for repositories)
  - removeCache() called from prepareIndexer() before it starts the real work

So I think w.r.t. conflicting with thread pools that's fine.

**Pull request merged**